### ISSUE TITLE
Feature request: Add uncertainty calculation using dropout during prediction

### ISSUE THREAD

**Issue by JZuegg** [deepchem/deepchem#1119]
As mentioned in the Lobby, I will start to implement a uncertainty calculation based on dropouts during prediction and collecting the variation of the prediction (see https://towardsdatascience.com/adding-uncertainty-to-deep-learning-ecc2401f2013).  Will post it here, but any comments, suggestions, help are welcome.

**Comment by patrickhop** [comment]
repost from gitter: http://mlg.eng.cam.ac.uk/yarin/blog_3d801aa532c1ce.html

**Comment by peastman** [comment]
Here's a recent paper you might find useful: https://arxiv.org/abs/1703.04977.  Basically there are two types of uncertainty that need to be handled in different ways: aleatoric uncertainty and epistemic uncertainty.  Fortunately, there are easy ways to get estimates of both.  Dropout is a way of measuring epistemic uncertainty.  To measure aleatoric uncertainty, you add a term to the loss function that requires it to estimate the uncertainty in every output value.

**Comment by JZuegg** [comment]
Quick request for technical coding feedback: 
Calculating the prediction using dropout can be set via the training parameter in tf. In your code this sits in TensorGraph._create_feed_dicts which is called by TensorGraph.predict_on_generator (fixed to training: False). 
[Q] What's better to extend the current  TensorGraph.predict_on_generator with a asTraining=False parameter (and corresponding  predict_* methods, and including a predict_proba_variance method, with n_post as number of predictions), or to overwrite them only in MultiTaskClassifier, the only model so far using dropout layers (?). The former would make it applicable to any TensorGraph with a dropout layer (would also set self.dropouts)...Not sure though if it upsets other parts of the code. Also depends on plans to add dropout layers to other models, and if tensorflow has other layers where 'training: true' has an effect.

**Comment by JZuegg** [comment]
Without being a coding expert, I was planning in the order of:

TensorGraph
def predicit_on_generator(self, generator, transformers=[], outputs=None, **training=False**):
	...
      for feed_dict in self._create_feed_dicts(generator, **training**):
        if self.queue_installed:
	...
	
def predict_proba(self, dataset, transformers=[], outputs=None, **training=False**):
    generator = self.default_generator(dataset, predict=True, pad_batches=False)
    return self.predicit_on_generator(generator, transformers, outputs, **training=training**)

def **predicit_proba_var**(self, dataset, transformers=[], outputs=None, n_post=100):
  training = True
  y_pred = []
  for _ in range(n_post):
    y_pred.append(self.predict_proba(dataset, transformers, outputs, **training**)
  pred_mean = np.mean(y_pred, axis=0)
  pred_var = np.var(y_pred, axis=0)
  return pred_var
	
Have something similar working (using separate methods), and testing it on a default MultiTaskClassifier (layer_size=[1000], dropout=0.5, feature=ECFP).

**Comment by peastman** [comment]
Let's give some thought to how we can design this in a way so that it's clear to the user what's happening, and that won't encourage them to do things they shouldn't.  Here are some of the important points.

1. There are multiple kinds of uncertainty.  Using dropout only measures epistemic uncertainty.  We either need to make that very clear to the user, or else consider ways of designing the feature so that it measures both epistemic and aleatoric uncertainty.  Otherwise, we're just going to give them a number, and they'll think that number means something different than it really does.

2. This method is only applicable to models that include dropout.  If a model doesn't have dropout, the method you outlined will still appear to work, but it will report the variance as 0, leading the user to draw totally wrong conclusions.  Even if a model does include dropout, it may only do so in a limited subset of layers.  The result will then only reflect the uncertainty coming from a subset of the model's calculations.  Dropout is an operation that can be used for multiple purposes: regularization and estimating uncertainty.  The two are related, but they're not identical.  For most models, the designer will have added Dropout layers where they thought it was important for regularization, but that might not be enough to give an accurate measurement of uncertainty.

So how do we design this to make sure a) the user only uses it when it's appropriate, b) we return a result that accurately measures what it's supposed to measure, and c) the user understands what that result means?

**Comment by JZuegg** [comment]
Ok. Was hoping that the toolkit is used for more experience user, but you're right.  

Naming the result carefully is important, hence, naming it a variance of prediction rather than uncertainty was my initial take. Also by using dropout layers, the variance will be a relative one, only valid within that model, and training and test set. My planned test case for the feature was to use the variance of the training set as reference population, and compare it against the variance population of a carefully selected test set, chemically diverse to the training set. I was also planning to use different (lower I guess) dropout rates for calculating the prediction variance, and see how they perform, following your argument of different application/purpose of dropouts.
  
Second, the above design was only a very quick implementation, but the current method can be implemented catching non-dropout models and return a None.... either by recording, on build, if model contains a dropout layer, or going through the layers and check for dropout (have to check if and how that can be done, ... any suggestions).

**Comment by rbharath** [comment]
Carrying over discussion from #1211: The uncertainty estimation is probably our most requested feature addition. Let's pick this thread up and see if we can make a gameplan for a reasonable implementation

**Comment by peastman** [comment]
Shall we perhaps make a list of different ways of computing uncertainty, then consider which ones we want to implement?  Here are the ones I know of.

1. Train and predict with dropout, as discussed above.
2. Train a collection of models and get a prediction from each one.
3. Add an output that contains an estimate of uncertainty and a corresponding loss term.

1 and 2 are comparable in the sense that they measure the same kind of uncertainty.  3 measures a fundamentally different kind of uncertainty.

**Comment by rbharath** [comment]
My guess is that 1/2 are what most people are looking for since they prove a form of confidence interval. 3 is pretty different and I think we'd be fine leaving it off from an initial implementation.

**Comment by peastman** [comment]
In a lot of cases, 3 is the more important factor.  Leaving it out could lead to really misleading results.

An easy way to understand the difference is to think of a linear model.  If your data is highly nonlinear, it will do a terrible job of fitting the data.  On the other hand, linear regression is deterministic.  If you run the fitting procedure several times, you'll get exactly the same result every time.  So method 2 will report that all your predictions have an uncertainty of 0, despite the fact that the model does a terrible job of predicting the training data!

It's the difference between, "How well does the data constrain the model?" and "How well does the model fit the data?"  Both are important and both contribute to the uncertainty in your predictions.

**Comment by peastman** [comment]
Certainly they're different from each other.  They're both based on the same idea though: create an ensemble of models, either by varying the dropout mask while holding the parameters fixed, or by varying the parameters while holding the dropout mask fixed.  Then you take the variation in their predictions as a measure of uncertainty.  Method 2 is more robust, but also very expensive.  Assuming you're training with dropout, method 1 is very cheap and makes a reasonable approximation.

I've been thinking about possible designs, and I think I'm converging on a solution.  This can't be implemented as a generic layer on top of the model.  It requires direct support from the model, since it needs to use dropout on every layer, include an extra output, and add two more terms to the loss.  It also needs to work differently for regression and classification.  So what I have in mind is a set of architectural changes to TensorGraph, plus additions to each model class that we want to support it.

In terms of the public API, it will be very simple.  When you create a model, you add the flag `uncertainty=True` to the constructor.  For each `predict*()` method in TensorGraph, we'll add a corresponding `predict_with_uncertainty*()` method that has almost the same API, but it does the prediction many times with different dropout masks.  For each output it will return two tensors: the predicted value (averaged over all the masks), and an uncertainty (calculated with equation 9 of https://arxiv.org/abs/1703.04977).

@JZuegg is it ok if I make a first pass at this?  I want to put some thought into the TensorGraph changes needed to support it.  Once that's done, there will be a lot of work to do supporting this for all the different models and we can split that up.

**Comment by JZuegg** [comment]
@peastman fine with me. I am currently offline for another week.

**Comment by rbharath** [comment]
Closing since implemented in #1218
